### PR TITLE
Update to junit5 library

### PR DESCRIPTION
**Pull request by drtimwright**
Working PR for getting concordion ready for junit 5 integration.

### CODE REVIEW

File: settings.gradle [hunk 1]
```diff
@@ -0,0 +1 @@
+if (JavaVersion.current() >= JavaVersion.VERSION_1_8) include ':java8'
```
Comment (nigelcharman) [inline settings.gradle]: I can't see where the name of the module is set - is this part of it? I'd be keen for it to be named concordion-junit5 rather than concordion-java8.

Comment (tenwit) [inline settings.gradle]: It's currently inferred from the name of the directory, and reinforced in settings.gradle. We can change those, and/or add the project.name property in the subproject's build.Gradle.

**Comment by drtimwright** [comment]
@nigelcharman : I have reverted the commit for that move. Tried a couple of things, but it's git that's being the problem. Basically, git records a move as a delete and a create and then tries to 'reconstruct' the move when required. The reconstruction isn't working for many of our files. But I'll keep trying...

**Comment by drtimwright** [comment]
Think I figured it out - use 'git mv' instead of 'mv'!

**Comment by nigelcharman** [comment]
Thanks Tim, much better! We might want to squash/rebase when finalising this change.

**Comment by nigelcharman** [comment]
From @drtimwright:

So now there are 3 main ways of integrating with JUnit 5 - I still need to check listeners are working, before example, after example, etc, etc:

* Annotate your class with `@ConcordionJUnit5` and our test engine does the heavy lifting (because there doesn't seem to be a nice way to fake a test method like we do for junit 4):

```
@ConcordionJUnit5
public class WithTestFactoryMethod {
}
```

* Have a `@TestFactory` method and junit5 runs the class

```
@ConcordionJUnit5
public class WithTestFactoryMethod {
    @TestFactory
    Iterable<DynamicTest> determineConcordionTests(Iterable<DynamicTest> concordionTests) {
        return concordionTests;
    }
}
```

* Have a field that we inject tests into and a `@TestFactory` method and junit5 runs the tests

```
@ConcordionJUnit5
public class WithFieldInjection {

    @ConcordionTests
    public Iterable<DynamicTest> concordionTests;

    @TestFactory
    Iterable<DynamicTest> determineConcordionTests() {
        return concordionTests;
    }
}
```

**Comment by jamesclark92** [comment]
Is this ever going to be merged?

**Comment by nigelcharman** [comment]
@jamesclark92 not in its current state and probably not this year. As mentioned above, there's multiple ways to integrate with JUnit and a bunch of checks we need to make with it. The recent changes to JUnit 5 also introduce new ways of integrating. We're keen to explore a number of these options and decide which one offers the best approach for Concordion.

While it feels prudent to have waited for JUnit 5 to mature more before committing ourselves, the reality is that all the committers have day jobs and simply haven't found the time to complete the integration. Personally, I won't have time to work on this until mid-November at the earliest.

**Comment by ssssteffff** [comment]
Hi, is this MR still planned, or is it dismissed?

Thanks

**Comment by nigelcharman** [comment]
Hi @ssssteffff, the functionality is still planned, see above https://github.com/concordion/concordion/pull/243#issuecomment-423804832.

The latest thinking is that we'll release 3.0.0 without this, but including some other enhancements that require Java 8. JUnit 5 support may follow in a 3.1.0 release, or later, depending when we get time to work on it.

Feel free to have a go at implementing it yourself and raise a PR if you get it working :)

**Comment by victordfb** [comment]
Hello! There are still plans to upgrade this in 2020?

**Comment by nigelcharman** [comment]
Hi @victordfb, at this point it's unlikely we'll get time to complete this in 2020. I'll catch up with @drtimwright soon to discuss.

Thanks for the additional context. Is JUnit4 no longer supported by Spring Boot?

**Comment by gicappa** [comment]
I would also need the JUnit 5 support since I'm using Quarkus.io and JUnit5 is the default choice. 
Do you have any idea on when this could be done? 
Can I give you an helping hand in any way?

**Comment by mwmahlberg** [comment]
I'd love to help. However, since this seems to be a rather extensive merge conflict, I might need some help with that ;)

**Comment by nigelcharman** [comment]
@gicappa Are you able to use the JUnit5 Vintage Engine with the JUnit4 ConcordionRunner? This has been the solution that has been followed by others who required JUnit5 support with Concordion. I can add instructions to the website if that works for you.

**Comment by nigelcharman** [comment]
@mwmahlberg and @gicappa thanks for the offers to help. 

While this PR gets a JUnit 5 runner working, it's not as clean as I would like - see https://github.com/concordion/concordion/pull/243/files#diff-505eefafcac0d02dd038bc3a5fd9cde1993b9a80ee48f4b7c1312d33dcb3a503R47 for example usage. 

I'm also not sure it would work with some of the features of JUnit 5. It would be great to have a runner that worked with JUnit5 extensions (see issue #245). We could possibly also deprecate the Concordion Parallel Run Extension in favour of JUnit5's [Parallel Execution](https://junit.org/junit5/docs/snapshot/user-guide/index.html#writing-tests-parallel-execution).

Some of the complexity of this PR is to allow Concordion to build with versions of Java prior to Java 8. With Concordion 3 now requiring Java 8 it should be somewhat simpler to integrate.

It's interesting to note that while Cucumber does have some support for JUnit 5:
`You will not find JUnit 5 advertised in other documentation yet. The tooling support for Junit Platform hasn't quite reached the point yet where I'd be comfortable changing that. And given that Cucumber JUnit works perfectly fine in combination with JUnit Vintage I see no reason too rush the upgrade either.` (https://github.com/cucumber/cucumber-jvm/issues/1149#issuecomment-611716745)

I imagine it would take days, if not weeks, to get a functional Concordion runner working with JUnit 5. Let me know if you have capacity to do this?

**Comment by gicappa** [comment]
> @gicappa Are you able to use the JUnit5 Vintage Engine with the JUnit4 ConcordionRunner? This has been the solution that has been followed by others who required JUnit5 support with Concordion. I can add instructions to the website if that works for you.

Yes, this is what I'm doing right now and it works.

**Comment by gicappa** [comment]
> @mwmahlberg and @gicappa thanks for the offers to help.
[...]
> I imagine it would take days, if not weeks, to get a functional Concordion runner working with JUnit 5. Let me know if you have capacity to do this?

I thought that if you have already a clear idea where to intervene to finalize the integration and you have no time to implement it, I could have helped following your indications. AFAIU this is not the case.

I'm afraid I don't have the time to study the code so deeply and to invest weeks, even if I really love the project. 
Thanks for your work!

**Comment by nigelcharman** [comment]
Hi @lorenzodee, one simple task that I haven't got round to is to document the workaround of using the JUnit5 Vintage Engine with the JUnit4 ConcordionRunner. This would be via a PR on the concordion/concordion-website repository. I'd suggest the documentation should go on the `Integrations` page.

If you have the time to look at the JUnit5 runner, I'd spend some time understanding the JUnit4 ConcordionRunner, reading the JUnit5 docs, especially the Advanced Topics and `TestEngine`s, and exploring Cucumber's JUnit5 engine. Then you should be in a position to try writing a Concordion test engine.
There may be some advantage to creating a multi-pass engine where the first pass discovers the test hierarchy and possibly creates a HierarchicalTestEngine.
If you decide to give it a go, let us know and shout out if you need any insights into the current code.

**Comment by lorenzodee** [comment]
Thanks @nigelcharman. So, is https://github.com/concordion/concordion/pull/243#issuecomment-326778994 still applicable?

From the hint of implementing `TestEngine` and/or extending `HierarchicalTestEngine`, I take it that using https://github.com/concordion/concordion/pull/243#issuecomment-326778994 has too much boilerplate, right?

I'm having trouble building the `master` branch. I'm getting an error with `org.ajoberstar.git-publish` dependency (see https://github.com/concordion/concordion/blob/7f275a860b10143886c940a81a6ed07eb9af046f/build.gradle). Is this normal? I had to adjust it from version `3.0.0` to `3.0.1`.

**Comment by nigelcharman** [comment]
Hi @lorenzodee, yes I'm keen to reduce the boilerplate if possible. It would be good to take a fresh look at JUnit5 and how other runners, such as Cucumber, work. The initial investigation was in the early days of JUnit 5, before there were other runners, and JUnit 5 has evolved since then. We may be able to utilise parts of Tim's solution once the solution space has been explored further.

I'm seeing the same with the grgit plugin. Would you raise a PR for the version update please?

**Comment by lorenzodee** [comment]
Please see PR #343

**Comment by nigelcharman** [comment]
> Please see PR #343

Thanks, I've also just updated the build to work with JDK 17 (as well as JDK 8 and 11) and to only publish the specifications on the `main` branch.

**Comment by lorenzodee** [comment]
While I'm _slowly_ wrapping my head around JUnit Platform's `TestEngine`, `HierarchicalTestEngine<T>`, `TestDescriptor`, and `Node<T>`, I've got some questions:

1. What's a good way to run a specification/example? I'm currently trying something like this:

```java
    FixtureRunner runner = getFixtureRunner(); // create runner from fixture class
    ResultSummary result = runner.run(null, getFixtureInstance());
    // Or, runner.run(exampleName, getFixtureInstance());
    result.assertIsSatisfied(getFixtureType());
```

2. I've noticed that some specifications simply run _nested_ specifications (i.e. act as containers). After running a specification (container), how are the results of the _nested_ specifications programmatically listed? I'll need this to help with reporting the number of tests that succeeded and failed.

3. How will Concordion fixtures be _marked_ to be discovered by the new JUnit Platform `TestEngine`?

    Currently, the JUnit Vintage engine detects the fixtures (annotated with `@RunWith(ConcordionRunner.class)`) as JUnit 4 tests.

    For the new/custom JUnit `TestEngine`, it could be the _absence_ of the `@RunWith` annotation _and_ the Java class name ending in either `Fixture` or `Test`. Or, it could be a new annotation (e.g. `@org.concordion.api.Fixture`). Any thoughts on this?

BTW, I've tried to make the existing tests work with JUnit Vintage. Please see https://github.com/lorenzodee/concordion/tree/junit5.

Thanks in advance.

**Comment by lorenzodee** [comment]
I've started with a _rough_ JUnit Platform `TestEngine` for Concordion. Please see https://github.com/lorenzodee/concordion/tree/junit5-engine. The following would be of interest:

1. Create `ConcordionTestEngine` (the Concordion `TestEngine` for the JUnit Platform)
    - For each fixture it discovers, corresponding `TestDescriptor` instances are created.
2. `SpecificationDescriptor` and `ExampleDescriptor` (both extends `FixtureBasedTestDescriptor`)
    - Mimics `ConcordionRunner` in processing specifications
3. Added `META-INF/services/org.junit.platform.engine.TestEngine` with reference to `ConcordionTestEngine`

Let me know if I'm on the right track, and may be I can submit a PR.

**Comment by nigelcharman** [comment]
@lorenzodee thanks for your work on this!! 

With regards to your comments around Extension vs Engine. I'm not sure it's such an issue for users to have to change to use `@BeforeExample`/`@AfterExample`, and `@BeforeSpecification`/`@AfterSpecification` rather than `@BeforeEach`/`@AfterEach`, and `@BeforeAll`/`@AfterAll` given that they need to change from `@Before`/`@After` and `@BeforeClass`/`@AfterClass`. The goal of introducing `@BeforeExample`/`@AfterExample`, and `@BeforeSpecification`/`@AfterSpecification` was to isolate users from the changes happening to JUnit. We can possibly support users to make the change with an automated refactoring (structural search and replace in IntelliJ IDEA).

I'll be keen for other opinions on this ^^^^. Has anyone on this thread got any opinions on it?

I'm not sure `@Rule`/`@ClassRule` is that widely used by Concordion users, but users need to find some solutions to this to move to JUnit 5 anyway.

As you've spotted, we sometimes use empty "container" classes to run nested specifications. We've had a few requests to remove the necessity for these when they are effectively empty classes. If you do see a way to no longer require these or generate them on the fly, I'm sure it would be welcome :)

I'll take a deeper look and answer some of your questions at the weekend.

**Comment by nigelcharman** [comment]
> I'll be keen for other opinions on this ^^^^. Has anyone on this thread got any opinions on it?

I've just raised the question on our Google Group to see if we get any feedback on that channel - https://groups.google.com/g/concordion/c/UJlPH-F575w

**Comment by nigelcharman** [comment]
> Let me know if I'm on the right track, and may be I can start some documentation and submit a PR.

It looks to me like it's heading in the right direction. I'll be keen to hear if you think there's any loose ends you need to tidy up?

Once you're raised a PR, we'll update the full test suite to use it and try it with some of the extensions and sample projects to check for any regressions.

**Comment by tisoft** [comment]
I would prefer a TestEngine implementation. Making it a TestExtension feels like a awkward workaround for me. 😂

**Comment by lorenzodee** [comment]
@nigelcharman Please see PR #345 . Thanks

1. Added `ConcordionTestEngine` and supporting classes
2. Added `@ConcordionFixture` (also marked as `@Testable`)
3. Modified `DefaultConcordionRunner` (the run-command runner) to use JUnit Platform engines (over JUnit 4's JUnitCore)

**Comment by lorenzodee** [comment]
**TL;DR** For the specs to succeed in JUnit Vintage, _without_ changes, the `ConcordionRunner` needs to avoid/delay instantiating the fixture until it is needed for running.

BTW, I've tried to make the existing tests work with JUnit Vintage. Please see https://github.com/lorenzodee/concordion/tree/junit5.

---

After further investigation, the difference between JUnit 4 and JUnit Vintage causes some specs/fixtures to fail. Here's how each would process Concordion fixtures/specs. More specifically, classes annotated with `@RunWith(ConcordionRunner.class)` and with a name that ends with either `Test` or `Fixture`.

Here, I'll be taking (`spec.concordion.common.command.example`) `ExampleFixture` and `FixtureWithoutExampleInitialisationFixture` as examples.

### JUnit 4

In JUnit 4, for each test class it finds, it processes one _before_ proceeding to process the next one. More specifically, it'll instantiate a `ConcordionRunner` to process `ExampleFixture` _before_ instantiating another `ConcordionRunner` to process `FixtureWithoutExampleInitialisationFixture`.

- When processing `ExampleFixture`, a `ConcordionRunner` is created. It instantiates the fixture in its constructor. So, an instance of `ExampleFixture` is created. And, since its specification (`Example.html`) includes `<a c:run="concordion" href="FixtureWithoutExampleInitialisation.html">` (i.e. a run command), it'll also end up processing `FixtureWithoutExampleInitialisationFixture` via another instance of `ConcordionRunner`. An (1st ☝️) instance of `FixtureWithoutExampleInitialisationFixture` is created. The specs will _succeed_ ✔️ because, at this point in time, there is only one instance of `FixtureWithoutExampleInitialisationFixture`.
- When processing `FixtureWithoutExampleInitialisationFixture`, a similar thing happens. A `ConcordionRunner` instance is created, which will instantiate (a 2nd ✌️) `FixtureWithoutExampleInitialisationFixture` in its constructor. But when the test is run, it detects that the test has _already_ been run. So, it _does not_ run a 2nd time. It uses the cached results (_succeeded_ ✔️).

### JUnit Vintage

In JUnit Vintage, it works in two phases: discovery and execution.

- In the discovery phase, it discovers `ExampleFixture` and `FixtureWithoutExampleInitialisationFixture`. It instantiates the runner objects (since both are annotated with `@RunWith(ConcordionRunner.class)`). It instantiates one `ConcordionRunner` for `ExampleFixture`, and one `ConcordionRunner` for `FixtureWithoutExampleInitialisationFixture` (i.e. one each). The `ConcordionRunner` instantiates the fixture in its constructor. This means that an (1st ☝️) instance of `FixtureWithoutExampleInitialisationFixture` is created.
- In the execution phase, it executes the discovered test classes: `ExampleFixture` and `FixtureWithoutExampleInitialisationFixture`. When the spec of `ExampleFixture` is processed, because it includes `<a c:run="concordion" href="FixtureWithoutExampleInitialisation.html">` (i.e. a run command), it'll also end up processing `FixtureWithoutExampleInitialisationFixture`. An (2nd ✌️) instance of `FixtureWithoutExampleInitialisationFixture` is created. The specs will _fail_ ❌ because, at this point in time, there are two instances of `FixtureWithoutExampleInitialisationFixture`. When the spec of `FixtureWithoutExampleInitialisationFixture` is processed, it detects that the test has _already_ been run. So, it _does not_ run a 2nd time. It uses the cached results (_failed_ ❌).

**Comment by nigelcharman** [comment]
I'm closing this PR with the work being done on #345 and #346 .

General conversation on JUnit 5 changes should be made on #141.
